server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,443 @@
+from gettext import gettext as _
+import logging
+import os
+import time
+import uuid
+import shutil
+
+import kobo.shortcuts
+
+import mongoengine
+from pulp.common import dateutils
+from pulp.plugins.util.publish_step import PublishStep
+from pulp.server.config import config as pulp_config
+from pulp.server.exceptions import PulpCodedException
+from pulp.server.db.model.repository import RepoPublishResult
+from pulp.server.db.model import Distributor
+
+
+ASSOCIATED_UNIT_DATE_KEYWORD = "created"
+START_DATE_KEYWORD = 'start_date'
+END_DATE_KEYWORD = 'end_date'
+
+_logger = logging.getLogger(__name__)
+
+
+class RSyncPublishStep(PublishStep):
+    _CMD = "rsync"
+    # recursive, symlinks, timestamps, keep dir links, omit dir times, compress, itemize
+    _AZQ = "-rtKOzi"
+
+    def __init__(self, step_type, file_list, src_directory, dest_directory, config=None,
+                 fast_forward=None, exclude=[], delete=False, links=False):
+        """
+        Copies files using rsync to a remote server and filesystem location defined in the
+        distributor config.
+
+        :param step_type: Name of step
+        :type step_type: str
+        :param file_list: list of paths relative to src_directory
+        :type file_list: list
+        :param src_directory: absolute path to directory which contains all items in file_list
+        :type src_directory: str
+        :param dest_directory: path to directory relative to remote root where files should be
+               rsynced
+        :type dest_directory: str
+        :param config: distributor configuration
+        :param fast_forward:
```

Comment:
The instance variable that is set by `fast_forward` parameter is not used anywhere. I am removing it.

This step simply rsyncs files that have already been selected by a step that is aware of this being an incremental publish.